### PR TITLE
SWC-6247: Add TrashCanList component

### PR DESCRIPTION
**Pull request by nickgros**
https://user-images.githubusercontent.com/17580037/181313144-6cb4d641-b37b-4997-9412-061fa7800d36.mov

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "allowSyntheticDefaultImports": true,
     "module": "esnext",
     "target": "es5",
-    "lib": ["es2019", "dom"],
+    "lib": ["es2020", "dom"],
```
Comment (nickgros) [inline tsconfig.json:9]: Need es2020 to use [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled)

File: src/lib/containers/trash/TrashCanList.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+          {item.originalParentId})
+        </>
+      </td>
+      <td>{formatDate(moment(item.deletedOn))}</td>
+      <td>
+        <Button size="sm" variant="outline" onClick={onRestore}>
+          Restore
+        </Button>
+      </td>
+      {/* <td>DeleteButtonPlaceholder</td> */}
```
Comment (nickgros) [inline src/lib/containers/trash/TrashCanList.tsx:53]:
```suggestion
```

File: src/lib/containers/trash/TrashCanList.tsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                    onChange={onSelectAll}
+                  />
+                </th>
+                <th>ID</th>
+                <th>Name</th>
+                {/* <th>TypePlaceholder</th> */}
+                <th>Location</th>
+                <th>Deleted On</th>
+                <th></th>
+                {/* <th>DeleteButtonPlaceholder</th> */}
```
Comment (nickgros) [inline src/lib/containers/trash/TrashCanList.tsx:159]:
```suggestion
```

File: src/lib/utils/hooks/SynapseAPI/trash/useTrashCan.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+) {
+  const queryClient = useQueryClient()
+  const { accessToken } = useSynapseContext()
+
+  return useMutation<
+    PromiseSettledResult<void>[],
+    SynapseClientError,
+    string | Set<string>
+  >(
+    (ids: string | Set<string>) => {
```
Comment (nickgros) [inline src/lib/utils/hooks/SynapseAPI/trash/useTrashCan.ts:62]: Accept one ID or a Set of IDs

File: src/lib/utils/hooks/SynapseAPI/trash/useTrashCan.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      if (typeof ids === 'string') {
+        ids = new Set([ids])
+      }
+      const promises = Array.from(ids).map(id => {
+        return SynapseClient.restoreFromTrashCan(id, accessToken)
+      })
+      return Promise.allSettled(promises)
```
Comment (nickgros) [inline src/lib/utils/hooks/SynapseAPI/trash/useTrashCan.ts:69]: Promise.allSettled returns an array of [PromiseSettledResult](https://github.com/microsoft/TypeScript/blob/8d0c72d5e03366a7c3ed710736dbde3abed5bf63/src/lib/es2020.promise.d.ts#L11).

This lets us handle each failure individually. In contrast, Promise.all will short-circuit and reject if one promise rejects.

File: src/lib/utils/SynapseClient.ts [hunk 6]
```diff
@@ -3771,3 +3775,42 @@ export function getProfilePicPreviewPresignedUrl(userId: string) {
     ),
   )
 }
+
+export function getItemsInTrashCan(
+  accessToken: string | undefined,
+  offset = 0,
+  limit = 2,
```
Comment (nickgros) [inline src/lib/utils/SynapseClient.ts:3782]:
```suggestion
  limit = 25,
```

**Review by jay-hodgson — APPROVED** [review]

File: tsconfig.json (hunk 1 above)
Comment (jay-hodgson) [reply tsconfig.json:9]: Travis build is still unhappy for some reason.
TS2550: Property 'allSettled' does not exist on type 'PromiseConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2020' or later.
105       return Promise.allSettled(promises)

Comment (nickgros) [reply tsconfig.json]: Needed to update tsconfig.build.json

**Pull request merged**